// otg-api/lib/otg_const.js
var otgTable = {
	// OpenTherm Gateway operating modes
	GatewayMode: {
		'0': { en: "Monitor" },
		'1': { en: "Gateway" }
	},
	// OpenTherm Gateway remote override normal or both
	RemoteOverride: { 
		'0': { en: "Low byte only", nl: "Nee, alleen lage byte" }, 
		'1': { en: "Low and high byte", nl: "Ja, in hoge en lage byte" }
	},
	// OpenTherm power levels
	PowerLevels: {
		'L': { en: "Low", nl: "Laag" },
		'M': { en: "Medium", nl: "Midden" },
		'H': { en: "High", nl: "Hoog" }
	},
	// OpenTherm Gateway LED functions
	LedFunction: {
		'R': { 
			en: "Receiving an OpenTherm message from the thermostat or boiler",
			nl: "Ontvangen van een OpenTherm boodschap door de thermostaat of ketel"
		},
		'X': {
			en: "Transmitting an OpenTherm message to the thermostat or boiler",
			nl: "Verzenden van een OpenTherm boodschap aan de thermostaat of ketel"
		},
		'T': {
			en: "Transmitting or receiving a message on the master interface",
			nl: "Verzenden of ontvangen van een OpenTherm boodschap op de master interface"
		},
		'B': {
			en: "Transmitting or receiving a message on the slave interface",
			nl: "Verzenden of ontvangen van een OpenTherm boodschap op de slave interface"
		},
		'O': {
			en: "Remote setpoint override is active",
			nl: "Doeltemperatuur wordt overschreven"
		},
		'F': {
			en: "Flame is on",
			nl: "Vlam is aan",
			var: "FlameStatus"
		},
		'H': {
			en: "Central heating is on",
			nl: "Verwarming is aan",
			var: "CHMode"
		},
		'W': {
			en: "Hot water is on",
			nl: "Tapwater is aan",
			var: "DHWMode"
		},
		'C': {
			en: "Comfort mode (Domestic Hot Water Enable) is on",
			nl: "Comfort mode (tapwater vraag) is aan",
			var: "DHWEnabled"
		},
		'E': {
			en: "Transmission error has been detected",
			nl: "Transmissiefout gedetecteerd",
			var: "Errors"
		},
		'M': {
			en: "Boiler requires maintenance",
			nl: "Ketel heeft onderhoud nodig",
			var: "DiagnosticEvent"
		},
		'P': {
			en: "Thermostat requests a raised power level",
			nl: "Thermostaat vraagt meer stroom"
		}
	},
	// OpenTherm Gateway GPIO configuration
	GPIOFunction: {
		   '0': { en: "None", nl: "Geen" },
		   '1': { en: "Ground (0V)" },
		   '2': { en: "Vcc (5V)" },
		   '3': { en: "LED E" },
		   '4': { en: "LED F" },
		   '5': { en: "Setback (low)", nl: "Setback (laag)" },
		   '6': { en: "Setback (high)", nl: "Setback (hoog)" },
		   '7': { en: "Temperature sensor", nl: "Temperatuur sensor" }
	},
	// OpenTherm Gateway signal transition checking
	IgnoreTrans: {
		'0': { en: "Check", nl: "Controleer" },
		'1': { en: "Ignore", nl: "Negeer" }
	},
	// OpenTherm Gateway reference voltage
	RefVoltage: {
		'0': { en: "0.625V" },
		'1': { en: "0.833V" },
		'2': { en: "1.042V" },
		'3': { en: "1.250V" },
		'4': { en: "1.458V" },
		'5': { en: "1.667V" },
		'6': { en: "1.875V" },
		'7': { en: "2.083V" },
		'8': { en: "2.292V" },
		'9': { en: "2.500V" }
	},
	// OpenTherm Gateway domestic hot water setting
	DHWsetting: {
		'0': { en: "Off", nl: "Uit" },
		'1': { en: "On (comfort mode)", nl: "Aan (comfort mode)" },
		'A': { en: "Thermostat controlled", nl: "Volgt thermostaat" }
	},
	// OpenTherm Gateway message types
	Initiator: { 
		'A': "Answer", 'B': "Boiler", 'R': "Request", 'T': "Thermostat"
	},
	// OpenTherm message types
	MsgType: {
		'0': "Read-Data", '1': "Write-Data", '2': "Invalid-Data", '3': "-reserved-",
		'4': "Read-Ack", '5': "Write-Ack", '6': "Data-Invalid", '7': "Unknown-DataId" 
	 },
	 // Gateway configuration commands
    Config: { 
		VER: { 
			en: "Gateway firmware version", nl: "Gateway firmware versie",
			var: "FirmwareVersion", rep: "A", ret: "OpenTherm Gateway (.*)" 
		},
		GW:  { 
			en: "Operating mode", nl: "Bedrijfsmodus",
			tab: "GatewayMode", var: "GatewayMode", cmd: "GW", rep: "G", ret: "[0|1]"
		},
		LED: { 
			en: "Function LED <A>", nl: "Functie van LED <A>",
			tab: "LedFunction", var: "LEDFunctions", cmd: "L<A>", rep: "L", ret: "[R|X|T|B|O|F|H|W|C|E|M]+", cnt: 4
		},
		ITR: { 
			en: "Non-significant transitions", nl: "Onbelangrijke transities",
			tab: "IgnoreTrans", var: "IgnoreTransitions", cmd: "IT", rep: "T", ret: "[0|1]"
		},
		REF: { 
			en: "Reference voltage", nl: "Referentie voltage",
			tab: "RefVoltage", var: "ReferenceVoltage", cmd: "VR", rep: "V", ret: "\\d"
		},
		HOT: { 
			en: "Domestic hot water enable", nl: "Tapwater verwarmen",
			tab: "DHWsetting", var: "DHWSetting", rep: "W", ret: "[0|1|A]"
		}
	},
	// Changes for firmware version 4
	VersionConfig: {
		'4': { 
			GW: { 
				rep: "M", ret: "PR: M=([M|G])", map: { 'M': "0", 'G': "1" } 
			},
			LED: {
				ret: "PR: L=([R|X|T|B|O|F|H|W|C|E|M|P]+)", cnt: 6
			},
			ITR: {
				en: "Ignore mid-bit transitions", nl: "Negeer transitites mid-bit",
				ret: "PR: T=([0|1])"
			},
			ROF: {
				en: "Remote Override Function flags", nl: "Remote Override Functie in beide bytes",
				tab: "RemoteOverride", var: "ROFInBothBytes", cmd: "OH", rep: "T", ret: "PR: T=[0|1]([0|1])"
			},
			REF: { 
				ret: "PR: V=(\\d)"
			},
			SETB: { 
				en: "Setback temperature", nl: "Setback temperatuur",
				var: "SetbackTemperature", 'cmd': "SB", 'rep': "S", 'ret': "PR: S=(\\d+.\\d+)" },
			HOT: {
				ret: "PR: W=([0|1|A])"
			},
			GPIO: { en: "GPIO configuration <A>", nl: "GPIO configuratie <A>",
				tab: "GPIOFunction", var: "GPIOConfiguration", cmd: "G<A>", rep: "G", ret: "PR: G=([\\d]+)", cnt: 2
			},
			PWR: { 
				en: "Current power level", nl: "Huidig stroomniveau",
				tab: "PowerLevels", var: "PowerLevel", rep: "P", ret: "PR: P=([L|M|H])"
			}
        }
	},
	// Valid OTG commands (per firmware 4.2.5)
	Commands: [ 
		'TT', 'TC', 'OT', 'SC', 'HW', 'PR', 'PS', 'GW', 'LA', 'LB', 'LC', 'LD', 'LE', 'LF', 'GA', 'GB',
		'SB', 'AA', 'DA', 'UI', 'KI', 'PM', 'SR', 'CR', 'SH', 'SW', 'MM', 'CS', 'CH', 'VS', 'RS', 'IT',
		'OH', 'FT', 'VR', 'DP'
	],
	Startup: [ "GW", "LED", "ITR", "ROF", "REF", "SETB", "HOT", "GPIO", "PWR" ]
};

module.exports = otgTable;
